---
<center><h1>Movie Recommendation System</h1></center>
<center><h3>Part of 30 Days 30 ML Projects Challenge</h3></center>

---

## 1) Understanding Problem Statement
---

In today's era of digital entertainment, the vast array of available movies and TV shows can overwhelm viewers when choosing what to watch. This project aims to tackle this issue through the development of a movie recommendation system, leveraging the power of data science and machine learning.

The problem can be classified as a **Recommendation System Machine Learning Problem**. The primary goal is **to construct a predictive model capable of suggesting personalized movie recommendations to users**. This model will analyze historical user preferences, movie ratings, and viewing habits to provide tailored movie suggestions. Additionally, it involves the application of **Collaborative Filtering**, **Content-Based Filtering**, or hybrid approaches to enhance recommendation accuracy.

By employing advanced recommendation algorithms and data analysis, this project seeks to simplify the decision-making process for viewers, enriching their entertainment experience while simultaneously demonstrating the practical use of machine learning in content recommendation systems.

## 2) Understanding Data
---

The project uses **Movies Data** which contains several variables (independent variables) and the outcome variable or dependent variable.

## 3) Getting System Ready
---
Importing required libraries


In [1]:
import numpy as np
import pandas as pd

# for text data preprocessing
import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer

# for model buidling
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

### Downloading stop words for text preprocessing

In [2]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\manis\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [3]:
# printing the stopwords in English
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

## 4) Data Eyeballing
---

### Laoding Data

In [4]:
movies_data = pd.read_csv('Datasets/Day18_Movies_Data.csv') 

In [5]:
movies_data

index     budget                                    genres   
0         0  237000000  Action Adventure Fantasy Science Fiction  \
1         1  300000000                  Adventure Fantasy Action   
2         2  245000000                    Action Adventure Crime   
3         3  250000000               Action Crime Drama Thriller   
4         4  260000000          Action Adventure Science Fiction   
...     ...        ...                                       ...   
4798   4798     220000                     Action Crime Thriller   
4799   4799       9000                            Comedy Romance   
4800   4800          0             Comedy Drama Romance TV Movie   
4801   4801          0                                       NaN   
4802   4802          0                               Documentary   

                                               homepage      id   
0                           http://www.avatarmovie.com/   19995  \
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4798                                                NaN    9367   
4799                                                NaN   72766   
4800  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4801                        http://shanghaicalling.com/  126186   
4802                                                NaN   25975   

                                               keywords original_language   
0     culture clash future space war space colony so...                en  \
1     ocean drug abuse exotic island east india trad...                en   
2            spy based on novel secret agent sequel mi6                en   
3     dc comics crime fighter terrorist secret ident...                en   
4     based on novel mars medallion space travel pri...                en   
...                                                 ...               ...   
4798  united states\u2013mexico barrier legs arms pa...                es   
4799                                                NaN                en   
4800  date love at first sight narration investigati...                en   
4801                                                NaN                en   
4802               obsession camcorder crush dream girl                en   

                                original_title   
0                                       Avatar  \
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4798                               El Mariachi   
4799                                 Newlyweds   
4800                 Signed, Sealed, Delivered   
4801                          Shanghai Calling   
4802                         My Date with Drew   

                                               overview  popularity  ...   
0     In the 22nd century, a paraplegic Marine is di...  150.437577  ...  \
1     Captain Barbossa, long believed to be dead, ha...  139.082615  ...   
2     A cryptic message from Bond’s past sends him o...  107.376788  ...   
3     Following the death of District Attorney Harve...  112.312950  ...   
4     John Carter is a war-weary, former military ca...   43.926995  ...   
...                                                 ...         ...  ...   
4798  El Mariachi just wants to play his guitar and ...   14.269792  ...   
4799  A newlywed couple's honeymoon is upended by th...    0.642552  ...   
4800  "Signed, Sealed, Delivered" introduces a dedic...    1.444476  ...   
4801  When ambitious New York attorney Sam is sent t...    0.857008  ...   
4802  Ever since the s

In [6]:
print('The size of Dataframe is: ', movies_data.shape)
print('-'*100)
print('The Column Name, Record Count and Data Types are as follows: ')
movies_data.info()
print('-'*100)

The size of Dataframe is:  (4803, 24)
----------------------------------------------------------------------------------------------------
The Column Name, Record Count and Data Types are as follows: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   index                 4803 non-null   int64  
 1   budget                4803 non-null   int64  
 2   genres                4775 non-null   object 
 3   homepage              1712 non-null   object 
 4   id                    4803 non-null   int64  
 5   keywords              4391 non-null   object 
 6   original_language     4803 non-null   object 
 7   original_title        4803 non-null   object 
 8   overview              4800 non-null   object 
 9   popularity            4803 non-null   float64
 10  production_companies  4803 non-null   object 
 11  production_countries  48

In [7]:
# Defining numerical & categorical columns
numeric_features = [feature for feature in movies_data.columns if movies_data[feature].dtype != 'O']
categorical_features = [feature for feature in movies_data.columns if movies_data[feature].dtype == 'O']

# print columns
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 8 numerical features : ['index', 'budget', 'id', 'popularity', 'revenue', 'runtime', 'vote_average', 'vote_count']

We have 16 categorical features : ['genres', 'homepage', 'keywords', 'original_language', 'original_title', 'overview', 'production_companies', 'production_countries', 'release_date', 'spoken_languages', 'status', 'tagline', 'title', 'cast', 'crew', 'director']


In [8]:
print('Missing Value Presence in different columns of DataFrame are as follows : ')
print('-'*100)
total=movies_data.isnull().sum().sort_values(ascending=False)
percent=(movies_data.isnull().sum()/movies_data.isnull().count()*100).sort_values(ascending=False)
pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])

Missing Value Presence in different columns of DataFrame are as follows : 
----------------------------------------------------------------------------------------------------


Total    Percent
homepage               3091  64.355611
tagline                 844  17.572351
keywords                412   8.577972
cast                     43   0.895274
director                 30   0.624610
genres                   28   0.582969
overview                  3   0.062461
runtime                   2   0.041641
release_date              1   0.020820
popularity                0   0.000000
production_companies      0   0.000000
production_countries      0   0.000000
budget                    0   0.000000
revenue                   0   0.000000
original_title            0   0.000000
spoken_languages          0   0.000000
status                    0   0.000000
original_language         0   0.000000
title                     0   0.000000
vote_average              0   0.000000
vote_count                0   0.000000
id                        0   0.000000
crew                      0   0.000000
index                     0   0.000000

In [10]:
print('Summary Statistics of numerical features for DataFrame are as follows:')
print('-'*100)
movies_data.describe()

Summary Statistics of numerical features for DataFrame are as follows:
----------------------------------------------------------------------------------------------------


index        budget             id   popularity       revenue   
count  4803.000000  4.803000e+03    4803.000000  4803.000000  4.803000e+03  \
mean   2401.000000  2.904504e+07   57165.484281    21.492301  8.226064e+07   
std    1386.651002  4.072239e+07   88694.614033    31.816650  1.628571e+08   
min       0.000000  0.000000e+00       5.000000     0.000000  0.000000e+00   
25%    1200.500000  7.900000e+05    9014.500000     4.668070  0.000000e+00   
50%    2401.000000  1.500000e+07   14629.000000    12.921594  1.917000e+07   
75%    3601.500000  4.000000e+07   58610.500000    28.313505  9.291719e+07   
max    4802.000000  3.800000e+08  459488.000000   875.581305  2.787965e+09   

           runtime  vote_average    vote_count  
count  4801.000000   4803.000000   4803.000000  
mean    106.875859      6.092172    690.217989  
std      22.611935      1.194612   1234.585891  
min       0.000000      0.000000      0.000000  
25%      94.000000      5.600000     54.000000  
50%     103.000000      6.200000    235.000000  
75%     118.000000      6.800000    737.000000  
max     338.000000     10.000000  13752.000000

In [9]:
print('Summary Statistics of categorical features for DataFrame are as follows:')
print('-'*100)
movies_data.describe(include='object')

Summary Statistics of numerical features for DataFrame are as follows:
----------------------------------------------------------------------------------------------------


genres                           homepage          keywords   
count    4775                               1712              4391  \
unique   1168                               1691              4219   
top     Drama  http://www.missionimpossible.com/  independent film   
freq      370                                  4                55   

       original_language   original_title   
count               4803             4803  \
unique                37             4801   
top                   en  Out of the Blue   
freq                4505                2   

                                                 overview   
count                                                4800  \
unique                                               4800   
top     In the 22nd century, a paraplegic Marine is di...   
freq                                                    1   

       production_companies   
count                  4803  \
unique                 3697   
top                      []   
freq                    351   

                                     production_countries release_date   
count                                                4803         4802  \
unique                                                469         3280   
top     [{"iso_3166_1": "US", "name": "United States o...   2006-01-01   
freq                                                 2977           10   

                                spoken_languages    status   
count                                       4803      4803  \
unique                                       544         3   
top     [{"iso_639_1": "en", "name": "English"}]  Released   
freq                                        3171      4795   

                       tagline     title   
count                     3959      4803  \
unique                    3944      4800   
top     Based on a true story.  The Host   
freq                         3         2   

                                                     cast  crew   
count                                                4760  4803  \
unique                                               4741  4776   
top     William Shatner Leonard Nimoy DeForest Kelley ...    []   
freq                                                    6    28   

                director  
count               4773  
unique              2349  
top     Steven Spielberg  
freq                  27

## 5) Data Cleaning and Preprocessing
---

### Replace the null values with a null string (Only for demonstration, here it is not neede)

In [10]:
mail_data = mail_data.where((pd.notnull(mail_data)),'')

In [11]:
mail_data

Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

### Label Encoding

#### label spam mail as 0;  ham mail as 1;

In [12]:
mail_data.loc[mail_data['Category'] == 'spam', 'Category',] = 0
mail_data.loc[mail_data['Category'] == 'ham', 'Category',] = 1

In [13]:
mail_data

Category                                            Message
0           1  Go until jurong point, crazy.. Available only ...
1           1                      Ok lar... Joking wif u oni...
2           0  Free entry in 2 a wkly comp to win FA Cup fina...
3           1  U dun say so early hor... U c already then say...
4           1  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567        0  This is the 2nd time we have tried 2 contact u...
5568        1               Will ü b going to esplanade fr home?
5569        1  Pity, * was in mood for that. So...any other s...
5570        1  The guy did some bitching but I acted like i'd...
5571        1                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [14]:
mail_data['Category'].value_counts()

Category
1    4825
0     747
Name: count, dtype: int64

### Stemming

In [15]:
porter_stemmer = PorterStemmer()

In [16]:
def stemming(content):
    stemmed_content = re.sub('[^a-zA-Z]',' ',content)
    stemmed_content = stemmed_content.lower()
    stemmed_content = stemmed_content.split()
    stemmed_content = [porter_stemmer.stem(word) for word in stemmed_content if not word in stopwords.words('english')]
    stemmed_content = ' '.join(stemmed_content)
    return stemmed_content

In [17]:
mail_data['Message'] = mail_data['Message'].apply(stemming)

In [18]:
mail_data['Message']

0       go jurong point crazi avail bugi n great world...
1                                   ok lar joke wif u oni
2       free entri wkli comp win fa cup final tkt st m...
3                     u dun say earli hor u c alreadi say
4                    nah think goe usf live around though
                              ...                        
5567    nd time tri contact u u pound prize claim easi...
5568                                b go esplanad fr home
5569                                    piti mood suggest
5570    guy bitch act like interest buy someth els nex...
5571                                       rofl true name
Name: Message, Length: 5572, dtype: object

## 6) Model Building
---

### Creating Feature Matrix (Independent Variables) & Target Variable (Dependent Variable)

In [19]:
# separating the data and labels
X = mail_data['Message'] # Feature matrix
y = mail_data['Category'] # Target variable

In [20]:
X

0       go jurong point crazi avail bugi n great world...
1                                   ok lar joke wif u oni
2       free entri wkli comp win fa cup final tkt st m...
3                     u dun say earli hor u c alreadi say
4                    nah think goe usf live around though
                              ...                        
5567    nd time tri contact u u pound prize claim easi...
5568                                b go esplanad fr home
5569                                    piti mood suggest
5570    guy bitch act like interest buy someth els nex...
5571                                       rofl true name
Name: Message, Length: 5572, dtype: object

In [21]:
y

0       1
1       1
2       0
3       1
4       1
       ..
5567    0
5568    1
5569    1
5570    1
5571    1
Name: Category, Length: 5572, dtype: object

In [22]:
# convert Y_train and Y_test values as integers
y = y.astype('int')

### Feature Extraction

#### Transform the text data to feature vectors that can be used as input to the Logistic regression

In [23]:
vectorizer = TfidfVectorizer()

In [24]:
vectorizer.fit(X)

X = vectorizer.transform(X)

In [25]:
X

<5572x6296 sparse matrix of type '<class 'numpy.float64'>'
	with 45096 stored elements in Compressed Sparse Row format>

In [26]:
print(X)

  (0, 6135)	0.23616756554565888
  (0, 5957)	0.19460776670194488
  (0, 4091)	0.24055424511726686
  (0, 2932)	0.28506031120996994
  (0, 2827)	0.3522946643655987
  (0, 2245)	0.19460776670194488
  (0, 2208)	0.1649859743034801
  (0, 2171)	0.14066343975170745
  (0, 1169)	0.27282796669086984
  (0, 964)	0.29761995607435426
  (0, 738)	0.29761995607435426
  (0, 736)	0.33630333732147566
  (0, 379)	0.26350491969128115
  (0, 190)	0.3522946643655987
  (1, 6056)	0.44597659211687757
  (1, 3785)	0.564793662023427
  (1, 3760)	0.2809319560263009
  (1, 2960)	0.4218982744467187
  (1, 2794)	0.4745440766926726
  (2, 6101)	0.21369536090695063
  (2, 6067)	0.16011115093017092
  (2, 5695)	0.13727833879237866
  (2, 5536)	0.2476330040187214
  (2, 5420)	0.1320245012320154
  (2, 5131)	0.22058857181065877
  :	:
  (5567, 784)	0.15667410716389937
  (5567, 724)	0.2920652264491494
  (5568, 2457)	0.37457404553349233
  (5568, 2171)	0.29597505521175127
  (5568, 1996)	0.5740672391289212
  (5568, 1704)	0.6652366917601374
  (5

### Train-Test Split

In [27]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=45)

In [28]:
print(X.shape, X_train.shape, X_test.shape)

(5572, 6296) (4457, 6296) (1115, 6296)


In [29]:
print(y.shape, y_train.shape, y_test.shape)

(5572,) (4457,) (1115,)


### Model Comparison : Training & Evaluation

In [30]:
models = [LogisticRegression, SVC, DecisionTreeClassifier, RandomForestClassifier]
accuracy_scores = []
precision_scores = []
recall_scores = []
f1_scores = []

for model in models:
    classifier = model().fit(X_train, y_train)
    y_pred = classifier.predict(X_test)
    
    accuracy_scores.append(accuracy_score(y_test, y_pred))
    precision_scores.append(precision_score(y_test, y_pred))
    recall_scores.append(recall_score(y_test, y_pred))
    f1_scores.append(f1_score(y_test, y_pred))

In [31]:
classification_metrics_df = pd.DataFrame({
    "Model": ["Logistic Regression", "SVM", "Decision Tree", "Random Forest"],
    "Accuracy": accuracy_scores,
    "Precision": precision_scores,
    "Recall": recall_scores,
    "F1 Score": f1_scores
})

classification_metrics_df.set_index('Model', inplace=True)
classification_metrics_df

Accuracy  Precision    Recall  F1 Score
Model                                                       
Logistic Regression  0.953363   0.949803  0.998965  0.973764
SVM                  0.972197   0.968907  1.000000  0.984208
Decision Tree        0.965919   0.979339  0.981366  0.980352
Random Forest        0.974888   0.972782  0.998965  0.985700

### Inference

In the context of spam mail prediction, the performance metrics of four different models—Logistic Regression, SVM, Decision Tree, and Random Forest—have been evaluated. 

- Overall, all models exhibit high accuracy, with SVM and Random Forest leading the way, achieving accuracy scores of approximately 97.22% and 97.49%, respectively. This suggests that these models are effective at correctly classifying emails as spam or legitimate.

- When considering precision, recall, and F1 Score, the SVM and Random Forest models again excel, with precision scores exceeding 96.89% and recall scores reaching 100%, indicating minimal false positives and false negatives. Decision Tree also performs well in terms of precision and recall.

In summary, SVM and Random Forest models stand out as strong contenders for spam mail prediction, with high accuracy, precision, and recall scores. However, the choice of the best model may also depend on specific requirements such as computational efficiency and interpretability.